### PR TITLE
Feat: need one Trait to set Rollout strategy of Workload

### LINKED ISSUE #5255: [Feature] need one Trait to set Rollout strategy of Workload
**Is your feature request related to a problem? Please describe.**

need one Trait to set Rollout strategy of Workload

**Describe the solution you'd like**

create one new Trait to patch it for workload

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by StevenLeiZhang**
Signed-off-by: StevenLeiZhang <zhangleiic@163.com>

### Description of your changes

add one Trait to  set Rollout strategy of Workload

Fixes #5255 

I have:

- [ ] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

1. apply Application with this trait
```
cd vela-templates/definitions/usage-examples
kubectl delete -f application-with-update-strategy.yaml 

```
2. check the details of Deployment

```
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    app.oam.dev/last-applied-configuration: '{"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{"oam.dev/kubevela-version":"v1.6.5"},"labels":{"app.oam.dev/app-revision-hash":"1066426d9c209bf5","app.oam.dev/appRevision":"application-with-update-strategy-v1","app.oam.dev/cluster":"","app.oam.dev/component":"helloworld","app.oam.dev/name":"application-with-update-strategy","app.oam.dev/namespace":"default","app.oam.dev/resourceType":"WORKLOAD","app.oam.dev/revision":"helloworld-v1","oam.dev/render-hash":"128a79095edf0299","workload.oam.dev/type":"webservice"},"name":"helloworld","namespace":"default"},"spec":{"replicas":5,"selector":{"matchLabels":{"app.oam.dev/component":"helloworld"}},"strategy":{"rollingUpdate":{"maxSurge":"20%","maxUnavailable":"30%"},"type":"RollingUpdate"},"template":{"metadata":{"labels":{"app.oam.dev/component":"helloworld","app.oam.dev/name":"application-with-update-strategy"}},"spec":{"containers":[{"image":"oamdev/hello-world:latest","name":"helloworld","ports":[{"containerPort":80,"name":"port-80","protocol":"TCP"}],"resources":{"limits":{"cpu":"0.5","memory":"1024Mi"},"requests":{"cpu":"0.5","memory":"1024Mi"}}}]}}}}'
    app.oam.dev/last-applied-time: "2023-01-03T13:47:09Z"
    deployment.kubernetes.io/revision: "1"
    oam.dev/kubevela-version: v1.6.5
  creationTimestamp: "2023-01-03T13:47:09Z"
  generation: 1
  labels:
    app.oam.dev/app-revision-hash: 1066426d9c209bf5
    app.oam.dev/appRevision: application-with-update-strategy-v1
    app.oam.dev/cluster: ""
    app.oam.dev/component: helloworld
    app.oam.dev/name: application-with-update-strategy
    app.oam.dev/namespace: default
    app.oam.dev/resourceType: WORKLOAD
    app.oam.dev/revision: helloworld-v1
    oam.dev/render-hash: 128a79095edf0299
    workload.oam.dev/type: webservice
  name: helloworld
  namespace: default
  resourceVersion: "1445431"
  uid: 3bc0b7e8-545c-4c26-a57b-7f7228d67f0b
spec:
  progressDeadlineSeconds: 600
  replicas: 5
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.oam.dev/component: helloworld
  strategy:
    rollingUpdate:
      maxSurge: 20%
      maxUnavailable: 30%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.oam.dev/component: helloworld
        app.oam.dev/name: application-with-update-strategy
    spec:
      containers:
      - image: oamdev/hello-world:latest
        imagePullPolicy: Always
        name: helloworld
        ports:
        - containerPort: 80
          name: port-80
          protocol: TCP
        resources:
          limits:
            cpu: 500m
            memory: 1Gi
          requests:
            cpu: 500m
            memory: 1Gi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
status:
  availableReplicas: 5
  conditions:
  - lastTransitionTime: "2023-01-03T13:47:23Z"
    lastUpdateTime: "2023-01-03T13:47:23Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2023-01-03T13:47:09Z"
    lastUpdateTime: "2023-01-03T13:47:26Z"
    message: ReplicaSet "helloworld-6dd6f5f8d6" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 1
  readyReplicas: 5
  replicas: 5
  updatedReplicas: 5
```

3. apply Daemon Application
```
cd vela-templates/definitions/usage-examples
kubectl apply -f application-daemon-with-update-strategy.yaml
```

4. check the DaemonSet
```
kubectl get daemonset node-exporter -o yaml


apiVersion: apps/v1
kind: DaemonSet
metadata:
  annotations:
    app.oam.dev/last-applied-configuration: '{"apiVersion":"apps/v1","kind":"DaemonSet","metadata":{"annotations":{"oam.dev/kubevela-version":"v1.6.5"},"labels":{"app.oam.dev/app-revision-hash":"1b60f7b669c5e8ee","app.oam.dev/appRevision":"application-node-exporter-v1","app.oam.dev/cluster":"","app.oam.dev/component":"node-exporter","app.oam.dev/name":"application-node-exporter","app.oam.dev/namespace":"default","app.oam.dev/resourceType":"WORKLOAD","app.oam.dev/revision":"node-exporter-v1","oam.dev/render-hash":"e8d9366dc035179d","workload.oam.dev/type":"daemon"},"name":"node-exporter","namespace":"default"},"spec":{"selector":{"matchLabels":{"app.oam.dev/component":"node-exporter"}},"template":{"metadata":{"labels":{"app.oam.dev/component":"node-exporter","app.oam.dev/name":"application-node-exporter"}},"spec":{"containers":[{"args":["--path.sysfs=/host/sys","--path.rootfs=/host/root","--no-collector.wifi","--no-collector.hwmon","--collector.filesystem.ignored-mount-points=^/(dev|proc|sys|var/lib/docker/.+|var/lib/kubelet/pods/.+)($|/)","--collector.netclass.ignored-devices=^(veth.*)$"],"image":"prom/node-exporter","imagePullPolicy":"IfNotPresent","name":"node-exporter","resources":{"limits":{"cpu":0.1,"memory":"250Mi"},"requests":{"cpu":0.1,"memory":"250Mi"}},"volumeMounts":[{"mountPath":"/host/sys","name":"sys"},{"mountPath":"/host/root","name":"root"}]}],"volumes":[{"hostPath":{"path":"/sys"},"name":"sys"},{"hostPath":{"path":"/"},"name":"root"}]}},"updateStrategy":{"rollingUpdate":{"maxSurge":"20%","maxUnavailable":"30%"},"type":"RollingUpdate"}}}'
    app.oam.dev/last-applied-time: "2023-01-03T13:49:55Z"
    deprecated.daemonset.template.generation: "1"
    oam.dev/kubevela-version: v1.6.5
  creationTimestamp: "2023-01-03T13:49:55Z"
  generation: 1
  labels:
    app.oam.dev/app-revision-hash: 1b60f7b669c5e8ee
    app.oam.dev/appRevision: application-node-exporter-v1
    app.oam.dev/cluster: ""
    app.oam.dev/component: node-exporter
    app.oam.dev/name: application-node-exporter
    app.oam.dev/namespace: default
    app.oam.dev/resourceType: WORKLOAD
    app.oam.dev/revision: node-exporter-v1
    oam.dev/render-hash: e8d9366dc035179d
    workload.oam.dev/type: daemon
  name: node-exporter
  namespace: default
  resourceVersion: "1446110"
  uid: e970d021-9793-4812-9c3a-d433253d357d
spec:
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.oam.dev/component: node-exporter
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.oam.dev/component: node-exporter
        app.oam.dev/name: application-node-exporter
    spec:
      containers:
      - args:
        - --path.sysfs=/host/sys
        - --path.rootfs=/host/root
        - --no-collector.wifi
        - --no-collector.hwmon
        - --collector.filesystem.ignored-mount-points=^/(dev|proc|sys|var/lib/docker/.+|var/lib/kubelet/pods/.+)($|/)
        - --collector.netclass.ignored-devices=^(veth.*)$
        image: prom/node-exporter
        imagePullPolicy: IfNotPresent
        name: node-exporter
        resources:
          limits:
            cpu: 100m
            memory: 250Mi
          requests:
            cpu: 100m
            memory: 250Mi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /host/sys
          name: sys
        - mountPath: /host/root
          name: root
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
      volumes:
      - hostPath:
          path: /sys
          type: ""
        name: sys
      - hostPath:
          path: /
          type: ""
        name: root
  updateStrategy:
    rollingUpdate:
      maxSurge: 0
      maxUnavailable: 30%
    type: RollingUpdate
status:
  currentNumberScheduled: 1
  desiredNumberScheduled: 1
  numberAvailable: 1
  numberMisscheduled: 0
  numberReady: 1
  observedGeneration: 1
  updatedNumberScheduled: 1

```

### CODE REVIEW

File: charts/vela-core/templates/defwithtemplate/update-strategy.yaml [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# Code generated by KubeVela templates. DO NOT EDIT. Please edit the original cue file.
+# Definition source cue file: vela-templates/definitions/internal/update-strategy.cue
+apiVersion: core.oam.dev/v1beta1
+kind: TraitDefinition
+metadata:
+  annotations:
+    definition.oam.dev/alias: ""
+    definition.oam.dev/description: Set update strategy for Deployment/DaemonSet/StatefulSet
+  name: update-strategy
```
Comment (wonderflow) [inline charts/vela-core/templates/defwithtemplate/update-strategy.yaml:9]:
```suggestion
  name: native-rollout
```

Comment (wonderflow) [reply charts/vela-core/templates/defwithtemplate/update-strategy.yaml:9]: `update-strategy` is a bit confusion while `native-rollout` is also not a good enough name, any other ideas? @StevenLeiZhang

Comment (StevenLeiZhang) [reply charts/vela-core/templates/defwithtemplate/update-strategy.yaml:9]: > `update-strategy` is a bit confusion while `native-rollout` is also not a good enough name, any other ideas? @StevenLeiZhang

what about k8s-update-strategy? It is similar to k8s-object and kruise-rollout, which indicates Provider and Intent/Target?

Comment (wonderflow) [reply charts/vela-core/templates/defwithtemplate/update-strategy.yaml:9]: @StevenLeiZhang OK

Comment (StevenLeiZhang) [reply charts/vela-core/templates/defwithtemplate/update-strategy.yaml]: > @StevenLeiZhang OK

updated

**Review by wonderflow — APPROVED** [review]
good job

**Pull request merged**